src/Chapter3.hs

```diff
@@ -480,6 +525,80 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
 
+data Building
+    = Church
+    | Library
+    deriving (Show)
+
+data Castle
+    = Nope
+    | CastleWithWalls String
+    | Castle String
+    deriving (Show)
+
+
+data House
+    = One
+    | Two
+    | Three
+    | Four
+    deriving (Show)
+
+
+data City = City
+    {cityBuilding :: Building
+    ,cityCastle   :: Castle
+    ,cityHouses   :: [House]
+    }
+    deriving (Show)
+
+
+-- buildCastle :: Castle -> City -> City
+-- buildCastle newCastle city = city {cityCastle = newCastle}
+
+{- | buildCastle — build a castle in the city. If the city already has a castle,
+   the old castle is destroyed, and the new castle with the __new name__ is built
+-}
+buildCastle :: String -> City -> City
+buildCastle name city@(City _ castle _) = case castle of
+    CastleWithWalls _ -> city {cityCastle = CastleWithWalls name}
+    Castle _          -> city {cityCastle = Castle name}
+    _                 -> city
+
+
+-- | buildHouse — add a new living house
+buildHouse :: City -> House -> City
+buildHouse c h = c {cityHouses = h : cityHouses c}
+
+
+
+{- | buildWalls — build walls in the city. But since building walls is a
+   complicated task, walls can be built only if the city has a castle
+   and at least 10 living __people__ inside in all houses of the city in total.
+-}
+buildWalls :: City -> City
+buildWalls city = case cityCastle city of
+    (Castle name) -> if houseCount >= 10
+                     then city {cityCastle = CastleWithWalls name}
+                     else city
+    _             -> city
+    where
+        houseCount = countHouses city
+
+houseToInt :: House -> Int
+houseToInt h = case h of
+    One   -> 1
+    Two   -> 2
+    Three -> 3
+    Four  -> 4
+
+countHouses :: City -> Int
+countHouses city = case cityHouses city of
+    []      -> 0
```

Comment:
`sum` returns `0` on empty lists, so no need to handle this case separately 🙂

Reply:
Ok I will update it.